src/write/str.rs

```diff
@@ -0,0 +1,182 @@
+use indexmap::IndexSet;
+use std::ops::{Deref, DerefMut};
+use vec::Vec;
+
+use common::DebugStrOffset;
+use write::{Result, Section, SectionId, Writer};
+
+/// An identifier for a string in a `StringTable.`
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct StringId(usize);
+
+/// A table of strings that will be stored in a `.debug_str` section.
+// Requirements:
+// - values are `[u8]`, null bytes are not allowed
+// - insertion returns a fixed id
+// - inserting a duplicate returns the id of the existing value
+// - able to convert an id to a section offset
+// Optional?
+// - able to get an existing value given an id
+//
+// Limitations of current implementation (using IndexSet):
+// - inserting a duplicate requires an allocation or a double lookup
+// - doesn't preserve offsets when updating an existing `.debug_str` section
+//
+// Possible changes:
+// - calculate offsets as we add values, and use that as the id.
+//   This would avoid the need for DebugStrOffsets but would make it
+//   hard to implement `get`.
+#[derive(Debug, Default)]
+pub struct StringTable {
+    strings: IndexSet<Vec<u8>>,
+}
+
+impl StringTable {
+    /// Add a string to the string table and return its id.
+    ///
+    /// `bytes` must not contain a null byte.
```

Comment:
Worth clarifying that this function will panic if `bytes` does contain a null byte.